### PR TITLE
RISC-V: Add basic support for RISC-V

### PR DESCRIPTION
**Pull request by dlrobertson**
### Description ###

Add basic RISCV support such that using `gef` with [`riscv-binutils-gdb`] and the example
shown for using [`riscv-isa-sim`] [Debugging with GDB] should work.

### Notes ###
  - As of GDB 8.1.1 RISCV is not supported. As a result, I have not updated the docs etc
to add RISCV to the officially supported architectures
  - This commit is essentially just what I could hack up last night and is far from a complete implementation

### How Has This Been Tested? ###

| Architecture | Yes/No                   | Comments               |
|--------------|:------------------------:|------------------------|
| x86-32       | :heavy_multiplication_x: | Not changed            |
| x86-64       | :heavy_multiplication_x: | Not changed            |
| ARM          | :heavy_multiplication_x: | Not changed            |
| AARCH64      | :heavy_multiplication_x: | Not changed            |
| MIPS         | :heavy_multiplication_x: | Not changed            |
| POWERPC      | :heavy_multiplication_x: | Not changed            |
| SPARC        | :heavy_multiplication_x: | Not changed            |

Used a few simple examples (similar to the example shown at [Debugging with GDB]) to
verify `is_branch_taken` returned the correct result etc.

### Checklist ###

- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required. (See comment in [Notes](#notes))
- [x] I have read and agree to the **CONTRIBUTING** document.

[Debugging with GDB]: https://github.com/riscv/riscv-isa-sim#debugging-with-gdb
[`riscv-isa-sim`]: https://github.com/riscv/riscv-isa-sim
[`riscv-binutils-gdb`]: https://github.com/riscv/riscv-binutils-gdb

### CODE REVIEW

**Comment by Grazfather** [comment]
Very nice. Will try to test this out tonight for you.

**Comment by Grazfather** [comment]
Spent some time trying to get the tools built, no luck so far. I could put in more time, but any chance you have a build script or a docker image with the tools setup so that I can test?

**Comment by dlrobertson** [comment]
> any chance you have a build script

[riscv-tools] bundles things up nicely.

>  or a docker image

The ubuntu packages list is listed on the [riscv-tools] README, so a standard ubuntu docker image with those installed should work.

[riscv-tools]: https://github.com/riscv/riscv-tools#riscv-tools-

**Comment by Grazfather** [comment]
Looks good!

```
->0x1001003e <main+62>        beqz   a5, 0x10010064 <main+100>  NOT taken [Reason: !(86!=0)]
```

```
   0x1001004a <main+74>        slli   t6, t6, 0x1d
->0x1001004e <main+78>        bltu   a1, a2, 0x10010010 <main+16>       TAKEN [Reason: 12<21]
   \->  0x10010010 <main+16>        addiw  a4, a4, -110
      0x10010014 <main+20>        andi   a4, a4, 255
```

File: gef.py [hunk 1]
```diff
@@ -2738,6 +2845,13 @@ def is_aarch64():
     return elf.e_machine == Elf.AARCH64
 
 
+@lru_cache()
+def is_riscv():
+    """Checks if `filename` is a AARCH64 ELF."""
```
Comment (Grazfather) [inline gef.py]: Oops I missed this: RISCV not AARCH

File: gef.py [hunk 2]
```diff
@@ -1386,6 +1387,112 @@ def all_registers_stripped(self):
...
+        raise OSError("Architecture {:s} not supported yet".format(cls.arch))
+
+    def is_conditional_branch(self, insn):
+        return insn.mnemonic.startswith("b")
+
+    def is_branch_taken(self, insn):
+        def convert_to_signed(x):
+            """Convert a python long value to its two's compliment"""
+            if is_elf32():
+                if (x & 0x80000000) != 0:
```
Comment (Grazfather) [inline gef.py]: `if x & 0x80000000:` is better.

File: gef.py [hunk 3]
```diff
@@ -1386,6 +1387,112 @@ def all_registers_stripped(self):
...
+
+    @classmethod
+    def mprotect_asm(cls, addr, size, perm):
+        raise OSError("Architecture {:s} not supported yet".format(cls.arch))
+
+    def is_conditional_branch(self, insn):
+        return insn.mnemonic.startswith("b")
+
+    def is_branch_taken(self, insn):
+        def convert_to_signed(x):
```
Comment (Grazfather) [inline gef.py]: `v` instead of x

**Comment by dlrobertson** [comment]
Np, I'll push a second PR with the fixes

**Pull request merged**